src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java

```diff
@@ -0,0 +1,364 @@
+import org.apache.http.conn.UnsupportedSchemeException;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.Logger;
+
+public class FileRetriever implements DataRetriever {
+
+	protected URI uri;
+	protected String destination;
+	protected Duration maxAge;
+	protected Duration timeout = Duration.ofSeconds(30);
+	protected int numRetries = 1;
+	protected String retrieverName;
+	protected Logger logger;
+	protected boolean passiveFTP = false;
+	
+	
+	public FileRetriever()
+	{
+		this(null);
+	}
+	
+	/**
+	 * This constructor takes a string that will be used to name the logging output file.
+	 * 
+	 * @param retrieverName - the name of this File Retriever - the log file produced by this class will be named "${retrieverName}.log"
+	 */
+	public FileRetriever(String retrieverName)
+	{
+		//logger.trace("Setting retrieverName to {}", retrieverName);
+		this.setRetrieverName(retrieverName);
+		this.logger = this.createLogger(this.retrieverName, "RollingRandomAccessFile", this.getClass().getName(), true, Level.DEBUG, this.logger, "Data Retriever");
+	}
+	
+	@Override
+	public void fetchData() throws Exception 
+	{
+		if (this.uri == null)
+		{
+			throw new RuntimeException("You must provide a URI from which the file will be downloaded!");
+		}
+		else if (this.destination.trim().length() == 0)
+		{
+			throw new RuntimeException("You must provide a destination to which the file will be downloaded!");
+		}
+		//Before fetching anything, we need to check to see if the file already exists.
+		Path pathToFile = Paths.get(this.destination);
+		if (Files.exists(pathToFile))
+		{
+			BasicFileAttributes attributes = Files.readAttributes(pathToFile, BasicFileAttributes.class);
+			
+			Instant fileCreateTime = attributes.creationTime().toInstant();
+			Instant now = Instant.now();
+			//If the file is older than the maxAge...
+			if (fileCreateTime.isBefore( now.minus(this.maxAge) ))
+			{
+				//TODO: Option to save back-ups of old files.
+				logger.debug("File {} is older than allowed amount ({}) so it will be downloaded again.",this.destination,this.maxAge);
+				downloadData();
+				logger.debug("Download is complete.");
+			}
+			else
+			{
+				logger.debug("File {} is not older than allowed amount ({}) so it will not be downloaded.",this.destination,this.maxAge);
+			}
+		}
+		else
+		{
+			logger.debug("File {} does not exist and must be downloaded.", pathToFile);
+			//if file does not exist, get it from the URL.
+			downloadData();
+		}
+		
+		// Print some basic file stats.
+		if (Files.exists(pathToFile))
+		{
+			if (Files.isReadable(pathToFile))
+			{
+				BasicFileAttributes attribs = Files.readAttributes(pathToFile, BasicFileAttributes.class);
+				logger.info("File Info: Name: {}, Size: {}, Created: {}, Modified: {}",this.destination,attribs.size(), attribs.creationTime(), attribs.lastModifiedTime());
```

Comment:
This is excellent because at some point in the future building a 'file download validator' for steps like addlinks and orthopairs could maybe use this